### PR TITLE
Add compile-address-map command

### PR DESCRIPTION
**Pull request by andywingo**
An address map is a file of the form

IPV4 {
  psid_length=foo
  shift=bar
}
...

Either of psid_length or shift can be given.  If neither is given, they
default to 0 and 16, respectively.

The idea is that we'll specify address maps apart from the binding
table.

### CODE REVIEW

**Comment by dpino** [comment]
Wouldn't be simpler to specify a Lua syntax compatible file that could be processed directly with `loadstring` or `dofile`? Something like:

```
"IPV4_1" = { 
  psid_length=foo,
  shift=bar
},
"IPV4_2" = { 
  psid_length=foo,
  shift=bar
},
```

`parse` opens the file, reads its content as a Lua table and converts it to the new format:

```lua
local function parse(file)
   local fd = io.open(file)
   local content = fd:read("*all")
   local t = loadstring(("return {%s}"):format(content))
   fd:close()

   local ret = {}
   for ipv4,psid_info in pairs(t) do
      table.insert(ret, convert_entry(ipv4, psid_info))
   end
   return ret
end
```
`convert_entry` would turn `ipv4` and `psid_info` params into:

```lua
{host_endian_ipv4=uint32, psid_length=value, shift=value}
```

I don't know, it's just an idea. Other configuration files in Snabb tend to be Lua tables too (ports configuration files in SnabbNFV come to my head [1]). Is there a powerful reason (performance maybe? but the file is processed only once, right?) to have a customized parser for a file with a custom syntax? That's my concern.

[1] https://github.com/Igalia/snabbswitch/blob/lwaftr_olive/src/program/snabbnfv/test_fixtures/nfvconfig/test_functions/same_vlan.ports

**Comment by andywingo** [comment]
Yeah I get what you're saying.  I wouldn't mind re-using Lua's parser if Lua had a way to parse a file without executing it.  However, with things as they are, I think using Lua's parser has a few down-sides:

 1. Lua's parser can't give us good source locations for domain-specific parse errors.  Like if what you type is bad Lua syntax, you get an OK error (though probably with a domain-specific parser we could make a better error).  But if you type good Lua syntax but bad syntax for the conf file, you can't later produce a source location for the domain-specific parse error.  With a custom parser we can do this.
 1. We are going to be having three configuration files, it seems (conf, binding table, address map), and it's easy to mix them up on the command line. Again, good parse errors can help users know what the error was.
 1. It could be that Lua syntax isn't actually what we want.  For example in your example, it's easy to forget the comma between the entries.  It's also easy to forget the quotes around the IP address.  Neither of these syntactic bits are necessary but using Lua's parse requires them.
 1. Of course, it is possible when using Lua's parser that loading a file produces some *other* error besides a parse error -- i.e. a runtime error, when evaluating the code.  This really bothers me, if you can't tell :)

Dunno.  I think going with our own grammar makes sense, if we can minimize the costs in terms of parser maintenance.  In that regard, I have another patch to clean up this one.  Let's see :)

**Comment by andywingo** [comment]
I was kinda thinking that we should be able to specify the address map using CIDR notation.  Like:

```
1.2.3.0/24 {psid_length=4}
```

To do this we probably need to tweak the RangeMap builder to accept ranges.  Will work on that.

**Comment by andywingo** [comment]
Alternately:

```
1.2.3.0-1.2.3.255 { psid_length=4 }
```

**Comment by andywingo** [comment]
Or even:
```
1.2.3.0-1.2.3.255, 4.5.6.7 { psid_length=4 }
```

On a practical level when we start simulating millions of entries in the binding table, it could be nice to have a readable address map at least.

**Comment by dpino** [comment]
@wingo: Agree with your comments. I think defining our own configuration syntax brings advantage. However, I was a bit worried about code maintenance of a new parser, specially if that happens for every configuration file we define. If the parser can be done with the less minimum code that would be the best solution for me.

I like the idea of ranges. If I have to pick one syntax over other I prefer the latter. Using CIDR makes me think there's a network address involved. The range makes explicit all the addresses within the range will have the same `psid_length` and `shift` values.

**Comment by andywingo** [comment]
Newest patch adds support for address ranges; see the README diff.  Tx for comments @dpino, would you mind taking another look please?  Thanks :)

**Comment by andywingo** [comment]
OK... thinking about this a little more, the compilation phase is pretty great but it's complicated to have to update a compiled address map.  The same will go for the binding table in the future.  My proposal is that:

 1. We add on a magic header to the on-disk format of compiled address maps and binding tables.  This lets us know if we are loading a compiled file or a textual file.
 2. The binding tables and address maps that the user specifies in the .conf file can be compiled or not.
 3. If the lwAFTR goes to load a compiled address map or binding table, that's great -- you have a real low-latency restart.
 4. If the lwAFTR realizes it's loading a textual address map or binding table, it will *automatically* try to look for a corresponding compiled version, by appending a ".o" extension to the input file name.
   1. If it finds such a file, it checks its header to make sure it's a compiled file of the right type.  The magic header also embeds the mtime of the textual file, with nanosecond resolution.  If the header has the right magic value and the mtime embedded in it matches the mtime of the textual file, sweet, we use the compiled file.  Otherwise we automatically compile the file, and attempt to replace the contents of the .o file.  In this case we print a message saying what it's doing.  If we fail to replace the .o file, we issue a warning.  We use the newly compiled file in any case.

This is similar to what Guile does with the Scheme files it runs.  The advantage is, the user keeps their textual mental model, while the lwAFTR retains the ability to start up fast, and we minimize the possibility that a compiled file does not correspond to a textual file.

This probably isn't an amazing win for address maps, but I think it's probably important for binding tables, given the perf numbers we see in PodHashMap's self-test.

**Comment by dpino** [comment]
@andywingo +1 on the proposed method for config file compilation and loading.

File: src/apps/lwaftr/address_map.lua [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+   struct { uint16_t psid_length; uint16_t shift; }
+]])
+
+local Parser = {}
+
+function Parser.new(file)
+   local name = '<unknown>'
+   if type(file) == 'string' then name, file = file, io.open(file) end
+   local ret = { column=0, line=0, name=name }
+   function ret.read_char() return file:read(1) end
```
Comment (dpino) [inline src/apps/lwaftr/address_map.lua]: I think it's necessary a method to close the file descriptor. I think a good point from where to call such method would be error, as it finalizes input reading from the file descriptor.

```lua
function ret.close() return file:close() end
```

```lua
function Parser:error(msg, ...)
   self:close()
   error(...)
end
```

File: src/program/snabb_lwaftr/compile_address_map/compile_address_map.lua [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+   return nil, err
+end
+
+function run(args)
+   local in_file, out_file = parse_args(args)
+   local map = address_map.compile(in_file)
+   if not out_file then
+      local base = in_file
+      print(base)
+      if base:match('[.]txt$') then base = base:sub(0, #base-4) end
```
Comment (dpino) [inline src/program/snabb_lwaftr/compile_address_map/compile_address_map.lua]: Nit: can use `gsub`:

```lua
out_file = base:gsub("%.txt$", "")
out_file = out_file .. ".map"
```

File: src/program/snabb_lwaftr/compile_address_map/compile_address_map.lua [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+   end
+   return nil, err
+end
+
+function run(args)
+   local in_file, out_file = parse_args(args)
+   local map = address_map.compile(in_file)
+   if not out_file then
+      local base = in_file
+      print(base)
```
Comment (dpino) [inline src/program/snabb_lwaftr/compile_address_map/compile_address_map.lua]: I'm not sure this is intended.

**Comment by dpino** [comment]
LGTM, except for comments.

**Comment by andywingo** [comment]
Fixed nits, going to see what travis does with this.  Thank you for the detailed review!

**Comment by andywingo** [comment]
Buf, I managed to merge this into wip-psid.  I think that was because I initially created the PR before wip-psid was merged, and wanted only the relevant diffs.  Bummer!  Will create a new one and merge it.

**Comment by andywingo** [comment]
Seems wip-psid isn't merged yet even, so I don't have to make a branch :) see #185.

**Pull request merged**